### ISSUE TITLE
[Features] Automatically retry requests on network errors

### ISSUE THREAD

**Issue by ousmanedev** [magicbell-io/magicbell-ruby#16]
Enable automatic retries on requests that fail due to a transient problem by configuring the maximum number of retries:

## Global configuration
```ruby
MagicBell.configure do |config|
  config.api_key = 'MAGICBELL_API_KEY'
  config.api_secret = 'MAGICBELL_API_SECRET'
  config.max_network_retries = 2
end
```

## Per-request configuration
```ruby
  magicbell = MagicBell::Client.new
  magicbell.max_network_retries = 2
```

## Request example
```ruby
# If the first request doesn't succeed, it will get retried twice

magicbell.create_notification(
  title: 'Rob assigned a task to you',
  recipients: [{
    external_id: 'DATABASE_ID'
  }]
)
```

**Comment by ousmanedev** [comment]
@josuemontano, any thought?

**Comment by unamashana** [comment]
@ousmanedev Sounds good. I'd include a default - maybe 2 retries?

**Comment by ecomba** [comment]
As I just started to look into the `magicbell-ruby` I saw that this issue has been open for quite a while. I was wondering if it would make sense to open a PR to add this new feature before I look into other things.

If so, I was wondering what constitutes a _**transient problem**_ in this scenario. Are we talking about timeouts, 5xx responses from the server, etc?

Is this something that would make sense to start with (before I jump into jucier bits) @unamashana ?

**Comment by unamashana** [comment]
@ecomba  Yes, this is a good one to start off with. You are right - transient would be 5xx, timeouts, etc. 

The idea comes from support for idempotent requests - https://www.magicbell.com/docs/rest-api/idempotent-requests
